Create an API key with `POST /api/api-key`

johnswanson opened the pull request:
This creates a new endpoint, `POST /api/api-key` that takes two arguments, a `group_id` and a `name`, and creates an
API key (along with a special API key user).

Fixes #36347

johnswanson committed:
Create an API key with `POST /api/api-key`

This creates a new endpoint, `POST /api/api-key` that takes two arguments, a `group_id` and a `name`, and creates an
API key (along with a special API key user).

I also went ahead and threw in the `api-key/count` endpoint here since
it's so small.

noahmoss commented on src/metabase/api/api_key.clj:
```diff
@@ -0,0 +1,75 @@
+(def ^:private user-first-name "API Key:")
+
+(defn- full-name [n]
+  (str user-first-name " " n))
```
I don't really see why the comparison on full name is needed? The only thing we need to enforce is that API key names are unique.

johnswanson replied:
I wanted to avoid the case where someone creates an API key with the same last name as a user and is blocked from doing so. But moot now since the refactor!

noahmoss commented on src/metabase/api/api_key.clj:
```diff
@@ -0,0 +1,75 @@
+      (let [user (first (t2/insert-returning-instances! :model/User
+                                                        {:email      email
+                                                         :password   (crypto-random/base64 16)
+                                                         :first_name "API Key:"
+                                                         :last_name  name
```
I just realized something that may require a change of approach here: `API Key` should almost certainly be translated, but it won't be if we're storing it as a raw string in the DB. 

Could you change it to be a i18n'ed string stored in memory that is hydrated on the `User` model when an internal user is fetched? 

Another schema adjustment that might make sense is to store the API key name in the `api_key` table and hydrate the `last_name` field when internal users are fetched, rather than storing it directly in the `core_user` table. This would solve the potential race condition below, right? Unless I'm misunderstanding something there.

i.e. an `api-key` user would just be a shell and all data related to the API key would be stored in the `api_key` table, instead of split across that and the `core_user` table

johnswanson replied:
I like this and implemented it! Thanks for the idea.

noahmoss commented on src/metabase/models/api_key.clj:
```diff
@@ -0,0 +1,52 @@
+(t2/define-before-insert :model/ApiKey
+  [{:keys [key key_prefix] :as api-key}]
+  (assert (= key_prefix (prefix key)))
+  (merge api-key
+         (hashed-key api-key)))
```
nit: I usually assume a function name describes what it returns, so I thought that `hashed-key` returned the key directly when skimming through this. Not a big deal either way but my bias would be for it to take and return strings, then do something like 
```suggestion
  (update api-key :key hashed-key))
```

johnswanson replied:
I like this better too, good call!

noahmoss reviewed (changes requested):
Looks pretty good overall, but I noticed an i18n issue which I think requires us to reevaluate some of the schema. Let me know if you have any questions about that (and sorry I didn't catch that earlier). Otherwise I just left some nits, but I'll do another pass whenever you're ready for a re-review

johnswanson committed:
Add a name to API keys

Instead of reusing the `core_user.last_name` column for the API key
name, it makes more sense to have an actual column that represents it.
For one thing, it allows us to have an actual UNIQUE index on the field
and assert that it's non-nullable at the database layer. It also just
seems better to not overload a column with multiple semantic types of
data.

johnswanson commented on src/metabase/api/api_key.clj:
```diff
@@ -0,0 +1,75 @@
+   [metabase.util.malli.schema :as ms]
+   [toucan2.core :as t2]))
+
+
```
Fixed!

johnswanson commented on src/metabase/api/api_key.clj:
```diff
@@ -0,0 +1,75 @@
+
+(def ^:private user-first-name "API Key:")
+
+(defn- full-name [n]
```
Removed the whole function, but good call :)

johnswanson commented on src/metabase/api/api_key.clj:
```diff
@@ -0,0 +1,75 @@
+  (:count
+   (t2/query-one {:select [[[:count :core_user.id] :count]]
+                  :from :core_user
+                  :where [:= (h2x/concat :first_name (h2x/literal " ") :last_name) (full-name n)]})))
```
Nice, I removed the function here but used this in the `GET /api-key/count` endpoint!

noahmoss commented on src/metabase/api/api_key.clj:
```diff
@@ -0,0 +1,57 @@
+     ;; duplicates. But a duplicate should be rare enough to just do multiple queries for now.
+     (if-not (t2/exists? :model/ApiKey :key_prefix prefix)
+       api-key
+       (throw (ex-info "could not generate key with unique prefix" {}))))))
```
If this message has a chance of being surfaced to end users it should be translated. (Granted, this one is extremely unlikely to be hit, but still...)

noahmoss commented on src/metabase/api/api_key.clj:
```diff
@@ -0,0 +1,57 @@
+                                                         :password (crypto-random/base64 16)
+                                                         :type     :api-key}))]
+        (user/set-permissions-groups! user [(perms-group/all-users) {:id group_id}])
+        (u/prog1 (-> (t2/insert-returning-instances! :model/ApiKey
```
nit: don't think `u/prog1` is needed anymore

noahmoss commented on src/metabase/models/api_key.clj:
```diff
@@ -0,0 +1,50 @@
+      add-prefix
+      (update :key u.password/hash-bcrypt)))
+
+(t2/define-before-update :model/ApiKey
+  [{:keys [key] :as api-key}]
+  (-> api-key
+      add-prefix
+      (update :key u.password/hash-bcrypt)))
```
If the `key` stored in the DB is already a hashed key, why do we need to do that again on updates? Wouldn't that hash it again and it would become corrupted?

(I'm assuming that if the key name is updated or something, the FE would include the hashed value as the `:key` since it should never have access to the unhashed value after it's generated)

johnswanson replied:
I'm actually planning for the `regenerate-key` endpoint here. We'll want to update and set a new `key` on an existing API key, so I was imagining we would do something like `(t2/update! :model/ApiKey {:id ...} {:key (generate-a-new-key)})`.

I was not imagining that someone would do `(t2/update! :model/ApiKey {:id ...} an-api-key-selected-from-the-db)`. That seems undesirable to me, since you'll risk race conditions by possibly overwriting values that were set by other threads - much better to only explicitly update the particular fields you're interested in updating.

But if this is a concern, one option might be to stop overloading `:key` to mean "hashed key" *and* the unhashed key. You could only write to the `unhashed_key` field, and only read from the `hashed_key` field. Might be better anyway - WDYT?

noahmoss replied:
Oh that's fair. But yeah it's a somewhat common pattern in the code base to read a model, change a single field, and then write it back, even if you're not planning on doing that directly here. I like the idea of not overloading the `:key` field to make it less error-prone

noahmoss commented on src/metabase/models/user.clj:
```diff
@@ -188,13 +188,23 @@
+  [{:keys [type] :as user}]
+  (cond-> user
+    (= :api-key type)
+    (assoc :first_name (i18n/tru "API Key:")
```
nit: our convention is usually to refer translation macros. `tru` is already available in this namespace without the `i18n` alias
```suggestion
    (assoc :first_name (tru "API Key:")
```

noahmoss replied:
Still need to be updated?

noahmoss commented on src/metabase/models/user.clj:
```diff
@@ -188,13 +188,23 @@
+  (cond-> user
+    (= :api-key type)
+    (assoc :first_name (i18n/tru "API Key:")
+           :last_name (:name (t2/select-one [:model/ApiKey :name] :user_id (:id user))))))
```
nit: marginally more concise (I think this is the syntax)
```suggestion
           :last_name (t2/select-one-fn :name :model/ApiKey :user_id (:id user)))))
```

noahmoss replied:
also maybe worth adding some error handling here in case the `api-key` user doesn't have an associated `ApiKey` model? shouldn't happen in practice but I don't think we enforce that as a DB invariant

johnswanson replied:
I tried to do this and it ended up breaking API key creation, because we need to:
- insert a new `:api-key`-typed user, grabbing the ID
- insert a new ApiKey with that `user_id`
Adding error-handling here causes this to break, because `t2/insert-returning-pk!` or `t2/insert-returning-instance!` both run after-select hooks.

Maybe we could set the `:last_name` to `(tru "[Missing]")` or something along those lines? Not ideal, but I'm not sure of a better option...

noahmoss replied:
Oh I see. I guess a workaround could be to avoid those functions when inserting the user and just use `t2/insert!`, and then `t2/query` to fetch the ID. But I don't think this error handling is _vital_, it's just nice to have.

cc @camsaul if there's a better t2 pattern to use in this sort of case; I don't really know the internals.

johnswanson committed:
CR response

johnswanson committed:
a much simpler solution...

johnswanson committed:
Fix failing test

A test is failing with inconsistent ordering when listing users. The
problem is that we order users by `first_name, last_name`, but API key
users have `null` values for both names, so there was no consistent
ordering. (Previously, it seems that we didn't have any `null`
first/last names in our tests.)

The test that caught this would be moot once we remove API key users
from the list of users we return from the `GET /users` endpoint, but the
bug would still be present, so it seems better to fix it by falling back
to `:id` to ensure consistent ordering.

noahmoss commented on src/metabase/models/api_key.clj:
```diff
@@ -0,0 +1,57 @@
+
+(defn- add-prefix [{:keys [unhashed_key] :as api-key}]
+  (cond-> api-key
+    key (assoc :key_prefix (prefix unhashed_key))))
```
I'm confused by this `cond->`. `key` is the condition, but it's a function so it'll always be truthy. Can you just use a simple `assoc` here without the `cond->`?

johnswanson replied:
Oops, this was a mistake from a rename. The condition should have been `unhashed_key` - we only want to attach a `key_prefix` if someone is trying to set the `unhashed_key`.

noahmoss commented on test/metabase/api/user_test.clj:
```diff
@@ -66,9 +66,10 @@
+          (is (= (disj (set (concat
+                             user/admin-or-self-visible-columns
+                             [:common_name :group_ids :personal_collection_id]))
+                       :type)
```
Would it be better to just exclude `:type` from `default-user-columns` if we're not typically going to return it when fetching users? removing `:type` manually here feels a little error-prone

johnswanson replied:
We needed to have it in `default-user-columns` so that the after-select hook would work - but now that we don't need to do any special massaging of the user's name, we can remove this.

noahmoss reviewed (approved):
Looks good overall, just a couple more nits

johnswanson committed:
Just store API key name in User.first_name

I'm still keeping around the `api_key.name` field. I think the lack of
normalization is worth the clean, DB-enforced name uniqueness.

johnswanson commented on src/metabase/api/api_key.clj:
```diff
@@ -0,0 +1,58 @@
+      (let [user (first (t2/insert-returning-instances! :model/User
+                                                        {:email    email
+                                                         :password (crypto-random/base64 16)
+                                                         :type     :api-key}))]
```
I saw that and was tempted, but the function is private and I wasn't sure whether it was worth making it public for this. WDYT?

noahmoss replied:
Oh you're right. I think that one would be fine to make public though.

noahmoss replied:
Lots of private functions, especially in model namespaces, are just private by default since they're not used outside of the namespace. But unless it's a namespace with a clearly marked public API, lots of functions are fine to make public if it's useful. This case definitely seems pretty benign to me.

johnswanson committed:
fix `cond->` for adding key_prefix to API key

noahmoss commented on test/metabase/api/api_key_test.clj:
```diff
@@ -0,0 +1,104 @@
+   [toucan2.tools.with-temp :as t2.with-temp]))
+
+(deftest api-key-creation-test
+  (t2.with-temp/with-temp
+      [:model/PermissionsGroup {group-id :id} {:name "Cool Friends"}]
```
nit: think this should be 2 spaces, not 4? same with the instances below.

johnswanson committed:
bugfix + use `user/insert-new-user!` to add user

I switched to using `user/insert-new-user!`, making that private, and
then noticed two subtleish bugs to fix while testing it:

1. I was threading `t2/insert-returning-instances!` (not the plural)
through `select-keys`, resulting in an empty map - so the resulting
response body lacked the `id`, `created_at`, and `updated_at` fields
from the API key itself. I don't know how I didn't notice this, I had a
test that asserted on the set of fields returned by the API endpoint,
but I made the assertion on the subset of fields that was wrong!

2. I was putting the raw `name`, passed by the user, into the fake,
invalid email address we construct for API keys. Characters invalid for
email address caused 500 errors. Use `u/slugify` before we construct the
email.

johnswanson committed:
fix spacing on `t2.with/temp` forms

I was curious why my editor was doing the wrong thing, so went to check
what the norm was in others' code - and found that I'm the only one who
wants to put the first argument to `t2.with-temp/with-temp` on its own
line. I'll just follow the local idiom here!
